### PR TITLE
[TTAHUB-534] add state filter control to filter panel

### PR DESCRIPTION
**Pull request by thewatermethod**
## Description of change
- added a "State" filter to the filter panel
- added a backend route that gets possible state values based on users regions -> grants

## How to test
- View the filter panel. Check out the state filter and enjoy how it filters by state. Also verify the correct states show in the panel (in my case, there were very few grants with states)

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-534

### Every PR

- [x] Meets issue criteria
- [x] JIRA ticket status updated
- [x] Code is meaningfully tested
- [x] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [n/a] API Documentation updated
- [n/a] Boundary diagram updated
- [n/a] Logical Data Model updated
- [n/a] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

**Comment by jasalisbury** [comment]
Whenever I click the state dropdown I don't have anything populated
Screenshot from 2021-12-15 13-50-30. Is that because I don't have any states in my database?

Also I see `- Selected -` as a disabled option in the dropdown menu of the filter and condition.

**Comment by thewatermethod** [comment]
> Whenever I click the state dropdown I don't have anything populated
> Is that because I don't have any states in my database?

The box should have for options all distinct stateCodes from the grants where grants->region.id == All regions you have permissions to. The reason I'm looking it up instead of just consulting some kind of dictionary is some region's users (11 & 12 I believe) can have disparate states, as far as I understand.

Also, I am far from a sequelize master so there could be an error there.

**Comment by thewatermethod** [comment]
> Also I see - Selected - as a disabled option in the dropdown menu of the filter and condition.

I was tweaking that, trying to remove some console warning based around using selected & value in a react component. I think I left off "hidden" when I was putting it back

**Comment by jasalisbury** [comment]
> > Whenever I click the state dropdown I don't have anything populated
> > Is that because I don't have any states in my database?
> 
> The box should have for options all distinct stateCodes from the grants where grants->region.id == All regions you have permissions to. The reason I'm looking it up instead of just consulting some kind of dictionary is some region's users (11 & 12 I believe) can have disparate states, as far as I understand.
> 
> Also, I am far from a sequelize master so there could be an error there.

We know (or have the ability to know) which states are in which regions. I _assume_ the regions don't provide TTA to grants outside their assigned states, but something to double check. If the list is static I could see us just having the list as a constant somewhere (`{ 1: 'NY', 'MA',....}`) so the state picker can just read the user's read regions (maybe combined with any region filter) and show relevant states to filter on. If the list is not static I could see it being useful to grab states based on what is in the DB.

**Comment by thewatermethod** [comment]
> > > Whenever I click the state dropdown I don't have anything populated
> > > Is that because I don't have any states in my database?
> > 
> > 
> > The box should have for options all distinct stateCodes from the grants where grants->region.id == All regions you have permissions to. The reason I'm looking it up instead of just consulting some kind of dictionary is some region's users (11 & 12 I believe) can have disparate states, as far as I understand.
> > Also, I am far from a sequelize master so there could be an error there.
> 
> We know (or have the ability to know) which states are in which regions. I _assume_ the regions don't provide TTA to grants outside their assigned states, but something to double check. If the list is static I could see us just having the list as a constant somewhere (`{ 1: 'NY', 'MA',....}`) so the state picker can just read the user's read regions (maybe combined with any region filter) and show relevant states to filter on. If the list is not static I could see it being useful to grab states based on what is in the DB.

The way it was explained to me is - regions 1-10, yes, we know what states they have. 11-12 vary. Doing it via an API call was my attempt to simplify it on our end (by which I mean, treat all users the same) but I can add an Object with a list of states for each of the known regions if you think that's better, and we can leave the API call in place for someone with one of the other two regions. Does that sound reasonable? Certainly an constant decreases weird errors I'd imagine.

**Comment by jasalisbury** [comment]
> > > > Whenever I click the state dropdown I don't have anything populated
> > > > Is that because I don't have any states in my database?
> > > 
> > > 
> > > The box should have for options all distinct stateCodes from the grants where grants->region.id == All regions you have permissions to. The reason I'm looking it up instead of just consulting some kind of dictionary is some region's users (11 & 12 I believe) can have disparate states, as far as I understand.
> > > Also, I am far from a sequelize master so there could be an error there.
> > 
> > 
> > We know (or have the ability to know) which states are in which regions. I _assume_ the regions don't provide TTA to grants outside their assigned states, but something to double check. If the list is static I could see us just having the list as a constant somewhere (`{ 1: 'NY', 'MA',....}`) so the state picker can just read the user's read regions (maybe combined with any region filter) and show relevant states to filter on. If the list is not static I could see it being useful to grab states based on what is in the DB.
> 
> The way it was explained to me is - regions 1-10, yes, we know what states they have. 11-12 vary. Doing it via an API call was my attempt to simplify it on our end (by which I mean, treat all users the same) but I can add an Object with a list of states for each of the known regions if you think that's better, and we can leave the API call in place for someone with one of the other two regions. Does that sound reasonable? Certainly an constant decreases weird errors I'd imagine.

Should 11 and 12 just list out every state?

**Comment by jasalisbury** [comment]
> > > > > Whenever I click the state dropdown I don't have anything populated
> > > > > Is that because I don't have any states in my database?
> > > > 
> > > > 
> > > > The box should have for options all distinct stateCodes from the grants where grants->region.id == All regions you have permissions to. The reason I'm looking it up instead of just consulting some kind of dictionary is some region's users (11 & 12 I believe) can have disparate states, as far as I understand.
> > > > Also, I am far from a sequelize master so there could be an error there.
> > > 
> > > 
> > > We know (or have the ability to know) which states are in which regions. I _assume_ the regions don't provide TTA to grants outside their assigned states, but something to double check. If the list is static I could see us just having the list as a constant somewhere (`{ 1: 'NY', 'MA',....}`) so the state picker can just read the user's read regions (maybe combined with any region filter) and show relevant states to filter on. If the list is not static I could see it being useful to grab states based on what is in the DB.
> > 
> > 
> > The way it was explained to me is - regions 1-10, yes, we know what states they have. 11-12 vary. Doing it via an API call was my attempt to simplify it on our end (by which I mean, treat all users the same) but I can add an Object with a list of states for each of the known regions if you think that's better, and we can leave the API call in place for someone with one of the other two regions. Does that sound reasonable? Certainly an constant decreases weird errors I'd imagine.
> 
> Should 11 and 12 just list out every state?

I guess we can check, if you list the states for 11 and 12 now how many do you get?

**Comment by thewatermethod** [comment]
> 

(locally)
SELECT "stateCode" from "Grants" WHERE ("regionId" = 11 or "regionId" = 12) AND "stateCode" is not null

MN, OK, NV, AZ, NM

**Comment by thewatermethod** [comment]
Patrice provided me with the link to the [region/state map](https://www.acf.hhs.gov/oro/regional-offices)

The jira specified "Region 11 and Region 12 should display states where they have grants," which is why I went down this road to begin with. That was decided at a meeting at some point. It's weird that most of my grants locally don't have state codes since I got fresh staging data this week

**Comment by jasalisbury** [comment]
> Patrice provided me with the link to the [region/state map](https://www.acf.hhs.gov/oro/regional-offices)
> 
> The jira specified "Region 11 and Region 12 should display states where they have grants," which is why I went down this road to begin with. That was decided at a meeting at some point. It's weird that most of my grants locally don't have state codes since I got fresh staging data this week

Cool, yeah grabbing them from the API seems appropriate.

File: frontend/src/components/filter/FilterStateSelect.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  const [stateCodes, setStateCodes] = useState([]);
+
+  // fetch state codes for user
+  useEffect(() => {
+    async function fetchStateCodes() {
+      try {
+        const codes = await getStateCodes();
+        setStateCodes(codes);
+      } catch (error) {
+        // fail silently
```
Comment (jasalisbury) [inline frontend/src/components/filter/FilterStateSelect.js:20]: We probably want to show some type of error if we can't retrieve the states. After think some more about this I wonder if the better solution is to just show all states for regions 11/12 and have a set list for each other region.

**Comment by kryswisnaskas** [comment]
Looks good! Works well. 👍 

A couple of comments.





> > Patrice provided me with the link to the [region/state map](https://www.acf.hhs.gov/oro/regional-offices)
> > The jira specified "Region 11 and Region 12 should display states where they have grants," which is why I went down this road to begin with. That was decided at a meeting at some point. It's weird that most of my grants locally don't have state codes since I got fresh staging data this week
> 
> Cool, yeah grabbing them from the API seems appropriate.

Could we get them from the API for regions 11 and 12 to fulfill the Jira requirement? (Unless Patrice changed her mind). I know there was a concern about processing errors. In that case, could we fall back to the list of all states?

BTW, the state codes have been populated now in production, and all of the non-prod environments, so the testing should be more meaningful.

I also noticed that the filter allows an invalid date, e.g. if the end date is greater than the start date. This results in the pill having an "invalid date" in it. This could be a separate PR as it seems unrelated to the current changes.

File: frontend/src/components/filter/FilterStateSelect.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
+import React, { useState, useEffect, useContext } from 'react';
+import PropTypes from 'prop-types';
+import FilterSelect from './FilterSelect';
+import UserContext from '../../UserContext';
+import { getStateCodes } from '../../fetchers/users';
+import { allRegionsUserHasPermissionTo } from '../../permissions';
+
+const ALL_STATES = [
```
Comment (jasalisbury) [inline frontend/src/components/filter/FilterStateSelect.js:8]: A comment explaining these are the states for regions in region order would be helpful.

File: src/routes/users/handlers.js [hunk 3]
```diff
@@ -21,3 +21,23 @@ export async function getPossibleCollaborators(req, res) {
     await handleErrors(req, res, error, { namespace: 'SERVICE:USER' });
   }
 }
+
+export async function getPossibleStateCodes(req, res) {
+  try {
+    const user = await userById(req.session.userId);
+    const regions = user.permissions.map((permission) => permission.regionId);
+    const grants = await Grant.findAll({
```
Comment (jasalisbury) [inline src/routes/users/handlers.js:29]: Can you move this to a method in the `services/grant.js` file? There are probably instances of us calling the DB from handlers but it is nice to keep DB calls inside services and outside of handlers

**Review by jasalisbury — COMMENTED** [review]
Some comments. The user I tested this on had `READ_WRITE` on all regions, `READ` on region 1, and `APPROVE` on region 1. The data is from the most recent production pull.
* State isn't showing up as a select-able filter on the activity reports page, should it?
* In the result from the API I see a "null" value
* _Most_ states are duplicated for me. Four that are not: DC, FC, FM, MH
* Do we want a "is not" state filter?

Also I'm not sure if "FC" is even a valid state/territory. I can't find it anywhere when googling.

**Comment by thewatermethod** [comment]
@kryswisnaskas 

> Could we get them from the API for regions 11 and 12 to fulfill the Jira requirement? (Unless Patrice changed her mind). I know there was a concern about processing errors. In that case, could we fall back to the list of all states?

Made this change

> I also noticed that the filter allows an invalid date, e.g. if the end date is greater than the start date. This results in the pill having an "invalid date" in it. This could be a separate PR as it seems unrelated to the current changes.

Should be fixed when #699 is merged in

**Review by jasalisbury — APPROVED** [review]
:ship:

**Pull request merged**